sources/softwareupdate.cpp

```diff
@@ -0,0 +1,239 @@
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with YIO-Remote software. If not, see <https://www.gnu.org/licenses/>.
+ *
+ * SPDX-License-Identifier: GPL-3.0-or-later
+ *****************************************************************************/
+
+#include "softwareupdate.h"
+
+static Q_LOGGING_CATEGORY(CLASS_LC, "softwareupdate");
+
+SoftwareUpdate *SoftwareUpdate::s_instance = nullptr;
+
+SoftwareUpdate::SoftwareUpdate(bool autoUpdate, QString appPath, BatteryFuelGauge *batteryFuelGauge, QObject *parent)
+    : QObject(parent), m_batteryFuelGauge(batteryFuelGauge), m_autoUpdate(autoUpdate), m_appPath(appPath) {
+    s_instance = this;
+
+    m_downloadTimer = new QElapsedTimer();
+
+    // start update checker timer
+    m_checkForUpdateTimer = new QTimer(this);
+    m_checkForUpdateTimer->setInterval(3600000);  // 1 hour
+    connect(m_checkForUpdateTimer, &QTimer::timeout, this, &SoftwareUpdate::onCheckForUpdateTimerTimeout);
+    m_checkForUpdateTimer->start();
+
+    qCDebug(CLASS_LC) << "Current version:" << m_currentVersion;
+
+    // check for update on startup as well
+    checkForUpdate();
+}
+
+SoftwareUpdate::~SoftwareUpdate() {
+    s_instance = nullptr;
+    m_checkForUpdateTimer->stop();
+    m_checkForUpdateTimer->deleteLater();
+}
+
+QObject *SoftwareUpdate::getQMLInstance(QQmlEngine *engine, QJSEngine *scriptEngine) {
+    Q_UNUSED(scriptEngine)
+    Q_ASSERT(s_instance);
+
+    QObject *instance = s_instance;
+    engine->setObjectOwnership(instance, QQmlEngine::CppOwnership);
+    return instance;
+}
+
+void SoftwareUpdate::onCheckForUpdateTimerTimeout() { checkForUpdate(); }
+
+void SoftwareUpdate::checkForUpdate() {
+    m_manager = new QNetworkAccessManager(this);
+    QNetworkRequest request;
+
+    connect(m_manager, &QNetworkAccessManager::finished, this, &SoftwareUpdate::checkForUpdateFinished);
+
+    request.setUrl(QUrl(m_updateUrl));
+    request.setRawHeader("Content-Type", "application/json");
+    request.setRawHeader("Accept", "application/json");
+
+    QJsonObject json;
+    json.insert("version", m_currentVersion);
+    QJsonDocument jsonDoc(json);
+    QByteArray    jsonData = jsonDoc.toJson();
+
+    m_manager->post(request, jsonData);
+}
+
+void SoftwareUpdate::checkForUpdateFinished(QNetworkReply *reply) {
+    if (reply->error() == QNetworkReply::NoError) {
+        QByteArray    responseData = reply->readAll();
+        QJsonDocument jsonResponse = QJsonDocument::fromJson(responseData);
+        QJsonObject   jsonObject   = jsonResponse.object();
+
+        if (jsonObject.contains("error")) {
+            qCWarning(CLASS_LC) << "Error" << jsonObject["error"].toString();
+            Notifications::getInstance()->add(true, tr("Software update: ") + jsonObject["error"].toString());
+        } else {
+            m_downloadUrl = jsonObject["url"].toString();
+            m_newVersion  = jsonObject["latest"].toString();
```

Comment:
Do we only publish releases or do we also allow a 'development channel', i.e. daily builds / snapshots?
Then we need to append a timestamp or build number to distinguish snapshot versions.
Might be could to already implement the optional suffix. But probably better to discuss this first...